src/dos/cdrom_ioctl_linux.cpp

```diff
@@ -113,4 +320,159 @@ bool CDROM_Interface_Ioctl::ReadSectorsHost([[maybe_unused]] void *buffer,
 	return false; /*TODO*/
 }
 
+void CDROM_Interface_Ioctl::InitAudio(int device_number)
+{
+	if (mixer_channel) {
+		return;
+	}
+
+	std::string name = std::string(MixerChannelNamePrefix) +
+	                   std::to_string(device_number);
+	audio_buffer.resize(AudioBufferBytes / 2);
+
+	auto callback = std::bind(&CDROM_Interface_Ioctl::CDAudioCallback,
+	                          this,
+	                          std::placeholders::_1);
+	mixer_channel = MIXER_AddChannel(callback,
+	                                 REDBOOK_PCM_FRAMES_PER_SECOND,
+	                                 name.c_str(),
+	                                 {ChannelFeature::Stereo,
+	                                  ChannelFeature::DigitalAudio});
+}
+
+void CDROM_Interface_Ioctl::CDAudioCallback(uint16_t requested_frames)
+{
+	if (audio_position == 0) {
+		if (sectors_remaining < 1) {
+			StopAudio();
+			return;
+		}
+		cdrom_read_audio cd = {};
+		cd.addr.lba         = current_sector;
+		cd.addr_format      = CDROM_LBA;
+		cd.buf = reinterpret_cast<uint8_t*>(audio_buffer.data());
+
+		// Maximum this ioctl can read in a single call is 1 second (75
+		// redbook frames)
+		cd.nframes = std::min(sectors_remaining, CD_FRAMES);
+		constexpr int PcmFramesPerRedbookFrame = BYTES_PER_RAW_REDBOOK_FRAME /
+		                                         BYTES_PER_REDBOOK_PCM_FRAME;
+		frames_in_buffer = cd.nframes * PcmFramesPerRedbookFrame;
+		current_sector += cd.nframes;
+		sectors_remaining -= cd.nframes;
+		if (ioctl(cdrom_fd, CDROMREADAUDIO, &cd) != 0) {
+			// Write out silence if something goes wrong
+			std::fill(audio_buffer.begin(), audio_buffer.end(), 0);
+
+#ifdef DEBUG_IOCTL
+			LOG_WARNING("CDROM_IOCTL: CDAudioCallback: CDROMREADAUDIO ioctl failed");
+#endif
+		}
+	}
+
+	const size_t frames_available = frames_in_buffer - audio_position;
+	const size_t frames_to_write = std::min(static_cast<size_t>(requested_frames),
+	                                        frames_available);
+
+	// audio_position must be doubled since a frame is two int16_t's for stereo sound
+#ifdef WORDS_BIGENDIAN
```

Comment:
Ok, forgot to mention this before... It gets all quite murky and confusing with the samples vs frames business. Some people also use the two interchangeably (not you, luckily 😅), which doesn't help either...

We introduced `AudioFrame` (`include/audio_frame.h`) for this very reason to simplify audio stream handling. I know exactly the distinction between frames and samples, but I managed to still confuse myself sometimes when doing buffer length manipulations, etc.; it's very easy to introduce bugs this way.

We have standardised on the use of `AudioFrames` in all audio rendering code that feeds the mixer. There are a few audio renderers that haven't been converted yet, but that's the ultimate goal. When that goal is accomplished, we will only keep `AddSamples_mfloat` and `AddSamples_sfloat`; every other `AddSample_*` variant will be nuked. That will simply `mixer.cpp` a lot too.

Some example on the use of `AudioFrame`:

#### gus.cpp
```cpp
using pan_scalars_array_t = std::array<AudioFrame, PAN_POSITIONS>;
```
#### gameblaster.cpp
```cpp
std::queue<AudioFrame> fifo = {};
```

####  opl.h
```cpp
std::queue<AudioFrame> fifo = {};
```

####  Misc signatures
```cpp
AudioFrame Process(const AudioFrame &in);
AudioFrame RenderFrame();
bool MaybeRenderFrame(AudioFrame &frame);
void Envelope::Process(const bool is_stereo, AudioFrame &frame)
```

The we can consistently use "frames" in our namings, and everything is crystal clear! For example:

```cpp
	while (frames_remaining) {
		if (float frame = 0.0f; MaybeRenderFrame(frame)) {
			channel->AddSamples_mfloat(1, &frame);
		}
		--frames_remaining;
	}
```

I should have said this earlier... Given our goals stated above, can I ask you to please migrate to using `AudioFrame` and use the term "frame" everywhere consistently? Thanks 😄

Reply:
That `AudioFrame` struct only stores floats (well it has a constructor for `int16_t` but it implicitly converts to float).

I'm a little confused how you want me to use this.  I've done a bit of audio coding in the past so I understand the value of wanting to do all your mixing in floats.  However, it's also my understanding you usually want to normalize your floats from -1.0f to 1.0f so I don't know if this cast will cut it.

Like you said earlier, CD audio is different from a SoundBlaster emulator or what have you in that I'm literally just reading `int16_t`'s off a disk.

Reply:
All good questions, and I'm here to help! Feel free to ask anything about audio if something is unclear.

Yes, using a `[-1.0f, 1.0f]` value range is common in audio processing, but that's just a convention—you can use any range you want, of course.

Currently, the "native" range of the mixer is `[INT16_MIN, INT16_MAX]` internally. You can see evidence of that here; the master compressor I introduced assumes `INT16_MIN` for the [0 dBFS](https://en.wikipedia.org/wiki/DBFS) level (more or less equivalent to "peak amplitude" in common language).
https://github.com/dosbox-staging/dosbox-staging/blob/main/src/hardware/mixer.cpp#L456

So eventually everything has to be normalised to `[INT16_MIN, INT16_MAX]` in our mixer, but you can configure the _expected_ sample value ranges per mixer channels individually with `SetPeakAmplitude()` and `Set0dbScalar()` (expected in the sense of what you pass into the channel with `AddSamples_*`):
https://github.com/dosbox-staging/dosbox-staging/blob/main/src/hardware/mixer.cpp#L840-L845
https://github.com/dosbox-staging/dosbox-staging/blob/main/src/hardware/mixer.cpp#L604-L614

You can read about the whole volume/peak scaling business here:
https://github.com/dosbox-staging/dosbox-staging/blob/main/include/mixer.h#L261-L295

By default, the peak of a mixer channel is `MAX_AUDIO`
https://github.com/dosbox-staging/dosbox-staging/blob/main/include/mixer.h#L291-L295

which is basically `INT16_MAX`
https://github.com/dosbox-staging/dosbox-staging/blob/main/include/mixer.h#L68

So the TL;DR is that yes, you can just cast the raw `int16_t` samples to `float` and simply use `Add_Samples_sfloat` and things will work. That way you will push samples into the mixer channel in our "native" `[INT16_MIN, INT16_MAX]` range. That's a lot of words for confiming "yes, just case to float", but I guess I wanted to explain it to you how it all fits together 😎 

Maybe it's a bit confusing and abstract, but check out how different audio devices set these mixer channel parameters, that should clear things up. However, you only need to cast to floats here.

Reply:
(And yes, probably I will refactor this all at some point so we use the more common `[-1.0f, 1.0f]` range. There are evolutionary reasons for normalising to the `int16_t` range, but I will rework that when I add 24-bit and 32-bit float audio capture support; then only the `[-1.0f, 1.0f]` internal range will make sense which can be written straight to disk for the 32-bit float capture).

Reply:
Interesting.  Yeah, if you get it normalized to -1.0f to 1.0f, you can also feed that directly to SDL (had a quick look and we're currently doing a conversion back to `int16_t` first).  It can accept floating point format but I believe it expects it to be normalized in that manner.

Most modern sound systems can accept float (I know PulseAudio in Linux will) but if not SDL will automatically convert it to what the platform needs.  I vaguely remember reading that Mac OSX **only** accepts floats so if that's true you end up going `int16_t->float->int16_t` then back to float again once SDL does the final conversion internally.

Reply:
Yeah, maybe 🤷🏻 0.00001% performance drop, I guess 😄 But it will be [-1.0f, 1.0f] anyway for the above outlined reasons in the future.

Reply:
Anyway, back to this PR.  I'm still not sure how you want me to use this `AudioFrame` struct.  I can do the conversion from `int16_t` to `float` in my callback function, no problem.  I think that would require two buffers (an `int_16` "read_buffer" to pass to the `ioctl` and then an `output_buffer` with float data to pass to the mixer.  But that's fine if that's what you want.

But then you said you wanted my buffer to be a vector of `AudioFrame`.  Problem is... the mixer doesn't accept that type.  It needs to be a vector of float.  I could `reinterpret_cast` and hope it works out (and it probably will until someone makes changes since the data is just 2 floats) but I think it's bordering on undefined behavior.